### PR TITLE
Omg 312 Challenging non-canonical IFEs using block txs

### PR DESCRIPTION
**Pull request by pdobacz**
polished version of scrapped https://github.com/omisego/elixir-omg/pull/366

don't review/merge before #361

This is still just logic, no integrations yet

### CODE REVIEW

File: apps/omg_watcher/lib/exit_processor/core.ex [hunk 1]
```diff
@@ -456,27 +613,53 @@ defmodule OMG.Watcher.ExitProcessor.Core do
...
+    end
   end
 
-  # tells whether a single transaction is a competitor for another single transactions, by returning nil or the
+  defp find_canonical(known_txs, raw_ife_tx) do
+    known_txs
+    |> Enum.find(fn %KnownTx{signed_tx: %Transaction.Signed{raw_tx: block_raw_tx}} -> block_raw_tx == raw_ife_tx end)
+    |> case do
+      nil -> {:error, :canonical_not_found}
+      got_it! -> {:ok, got_it!}
```
Comment (paulperegud) [inline apps/omg_watcher/lib/exit_processor/core.ex]: This exclamation mark is surprising. Why not use something plain and boring like `value` instead?

File: apps/omg_watcher/lib/exit_processor/request.ex [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+defmodule OMG.Watcher.ExitProcessor.Request do
+  @moduledoc """
+  Encapsulates the state of processing of `OMG.Watcher.ExitProcessor` pipelines
+
+  Holds all the necessary query date and the respective response
+
+  NOTE: this is highly experimental, to test out new patterns to follow when doing the Functional Core vs Imperative
+        Shell separation. **Do not yet** follow outside of here. I'm not sure whether such struct offers much and it
+        has its problems.
+  """
```
Comment (paulperegud) [inline apps/omg_watcher/lib/exit_processor/request.ex]: I really like this.

File: apps/omg_watcher/lib/exit_processor.ex [hunk 3]
```diff
@@ -208,23 +228,77 @@ defmodule OMG.Watcher.ExitProcessor do
...
+  end
+
+  defp run_utxo_exists(%ExitProcessor.Request{utxos_to_check: positions} = request) do
+    %{request | utxo_exists_result: positions |> Enum.map(&State.utxo_exists?/1)}
+  end
+
+  defp run_spend_getting(%ExitProcessor.Request{spends_to_get: positions} = request) do
+    %{
+      request
+      | spent_blknum_result: positions |> Enum.map(&Utxo.Position.to_db_key/1) |> Enum.map(&OMG.DB.spent_blknum/1)
```
Comment (paulperegud) [inline apps/omg_watcher/lib/exit_processor.ex]: maybe create a variable holding new `spent_blknum_result` value for better readability?

File: apps/omg_watcher/lib/exit_processor/core.ex [hunk 4]
```diff
@@ -52,10 +54,32 @@ defmodule OMG.Watcher.ExitProcessor.Core do
...
+    """
+    defstruct [:signed_tx, :utxo_pos]
+
+    @type t() :: %__MODULE__{
+            signed_tx: Transaction.Signed.t(),
+            utxo_pos: Utxo.Position.t()
+          }
+  end
+
+  alias __MODULE__.KnownTx
```
Comment (paulperegud) [inline apps/omg_watcher/lib/exit_processor/core.ex]: Is this line needed? It compiles correctly without it.

File: apps/omg_watcher/lib/exit_processor/core.ex [hunk 5]
```diff
@@ -485,7 +668,43 @@ defmodule OMG.Watcher.ExitProcessor.Core do
...
+    recovered
+  end
+
+  # based on an enumberable of `Utxo.Position` and a mapping that tells whether one exists it will pick
+  # only those that **were checked** and were missing
+  # (i.e. those not checked are assumed to be present)
+  defp only_utxos_checked_and_missing(utxo_positions, utxo_exists?) do
+    # the default value below is true, so that the assumption is that utxo not checked is **present**
+    utxo_positions
+    |> Enum.filter(fn pos -> !Map.get(utxo_exists?, pos, true) end)
```
Comment (paulperegud) [inline apps/omg_watcher/lib/exit_processor/core.ex]: This one can be more effectively done using `MapSet.difference/2`.

File: apps/omg_watcher/lib/exit_processor/core.ex (hunk 1 above)
Comment (pdobacz) [reply apps/omg_watcher/lib/exit_processor/core.ex]: T_T, oh okaaay

File: apps/omg_watcher/lib/exit_processor/core.ex (hunk 5 above)
Comment (pdobacz) [reply apps/omg_watcher/lib/exit_processor/core.ex]: I'll leave a TODO, but as discussed, there are much more expensive calls around here

**Pull request merged**